### PR TITLE
Lock/unlock tables when editing entries

### LINKED ISSUE #2472: When editing entries, Symphony will have time slots without data
When editing existing entries (see `EntryManager::edit`), Symphony will iterate over all fields, delete existing data first, then insert a new row for the field's data. So there will always be a small time slot without data.

This caused massive headaches for me the other day, because I experienced simultaneous processes which edited entry data, but every one of them relied on data being available. My solution was to stop using `EntryManager::edit`. Instead I added an additional unique key to the table and updated data using `Symphony::Database()->insert($data, 'tbl_entries_data_' . $field_id, true);` (where `true` will make Symphony add `ON DUPLICATE KEY UPDATE` to the query).

Still, after having applied this hackish fix, I asked myself: Why doesn't Symphony lock the table?

Like so, basically:
- `LOCK TABLES tbl_entries_data_{$field_id} WRITE`
- delete data
- insert data
- `UNLOCK TABLES`

Is there any reason against this?

### PR DESCRIPTION
**Pull request by michael-e**
When editing an entry, Symphony will iterate over all fields, delete existing data first, then insert a new row for the field's data. So there will always be a time slot without field data. If multiple processes edit the same entry, this means a race condition that may lead to data being corrupted in several ways (also depending on the field type). Locking the table solves the issue.

Fixes #2472

### CODE REVIEW

File: symphony/lib/toolkit/class.entrymanager.php [hunk 1]
```diff
@@ -174,39 +174,48 @@ public static function edit(Entry $entry)
...
+                    $fields[$index] = array_merge($data, $field_data);
+                }
+
+                Symphony::Database()->insert($fields, 'tbl_entries_data_' . $field_id);
+            } catch (Exception $e) {
+                Symphony::Log()->pushExceptionToLog($e, true);
             }
 
-            Symphony::Database()->insert($fields, 'tbl_entries_data_' . $field_id);
+            if ($did_lock) {
```
Comment (nitriques) [inline symphony/lib/toolkit/class.entrymanager.php]: This block *should* be in a finally statement to be 100% kasher (if I am not wrong?)

File: symphony/lib/toolkit/class.entrymanager.php [hunk 2]
```diff
@@ -174,39 +174,48 @@ public static function edit(Entry $entry)
                 continue;
             }
 
+            $did_lock = false;
             try {
+                $did_lock = Symphony::Database()->query('LOCK TABLES tbl_entries_data_' . $field_id . ' WRITE');
                 Symphony::Database()->delete('tbl_entries_data_' . $field_id, sprintf("
```
Comment (nitriques) [inline symphony/lib/toolkit/class.entrymanager.php]: The `WRITE` part means the table is still available for reads ?

**Comment by nitriques** [comment]
Thanks Michael. See my comments!

File: symphony/lib/toolkit/class.entrymanager.php (hunk 1 above)
Comment (michael-e) [inline symphony/lib/toolkit/class.entrymanager.php]: We discussed that some months ago (https://github.com/symphonycms/symphony-2/issues/2472#issuecomment-134315796); the issue is that `finally` requires PHP 5.5+.

File: symphony/lib/toolkit/class.entrymanager.php (hunk 2 above)
Comment (michael-e) [inline symphony/lib/toolkit/class.entrymanager.php]: No, it means that no other session can access the table, which is exactly what we need. See http://dev.mysql.com/doc/refman/5.7/en/lock-tables.html

Comment (nitriques) [inline symphony/lib/toolkit/class.entrymanager.php]: From what I am reading: 

> No other session can access it until the lock is released.

This is then an exclusive lock. Other operation are "queued". And yes, this is what we need ;)

File: symphony/lib/toolkit/class.entrymanager.php (hunk 1 above)
Comment (nitriques) [inline symphony/lib/toolkit/class.entrymanager.php]: I always forget about that. Thanks!

**Comment by nitriques** [comment]
I merge ASAP then!

**Pull request merged**